test/unit_test/runtime/magnitude_test.cpp

```diff
@@ -385,27 +262,79 @@ TEST_CASE("is_prime detects primes")
   }
 }
 
+TEST_CASE("is_valid_base_power")
+{
+  SECTION("0 power is invalid") {
+    REQUIRE(is_valid_base_power(base_power{2}));
+    CHECK(!is_valid_base_power(base_power{2, 0}));
+
+    REQUIRE(is_valid_base_power(base_power{41}));
+    CHECK(!is_valid_base_power(base_power{41, 0}));
+
+    REQUIRE(is_valid_base_power(base_power<pi_base>{}));
+    CHECK(!is_valid_base_power(base_power<pi_base>{0}));
+  }
+
+  SECTION("non-prime integers are invalid") {
+    CHECK(!is_valid_base_power(base_power{-8}));
+    CHECK(!is_valid_base_power(base_power{0}));
+    CHECK(!is_valid_base_power(base_power{1}));
+
+    CHECK(is_valid_base_power(base_power{2}));
+    CHECK(is_valid_base_power(base_power{3}));
+
+    CHECK(!is_valid_base_power(base_power{4}));
+  }
+
+  SECTION("non-positive floating point bases are invalid") {
+    CHECK(!is_valid_base_power(base_power<invalid_zero_base>{}));
+    CHECK(!is_valid_base_power(base_power<invalid_negative_base>{}));
+  }
+}
+
 TEST_CASE("pairwise_all evaluates all pairs")
 {
   SECTION("always true for empty tuples")
   {
-    CHECK(pairwise_all(std::make_tuple(), [](auto, auto){ return true; }));
-    CHECK(pairwise_all(std::make_tuple(), [](auto, auto){ return false; }));
+    CHECK(pairwise_all{[](auto, auto){ return true; }}());
+    CHECK(pairwise_all{[](auto, auto){ return false; }}());
   }
 
   SECTION("always true for single-element tuples")
   {
-    CHECK(pairwise_all(std::make_tuple(1), [](auto, auto){ return true; }));
-    CHECK(pairwise_all(std::make_tuple(3.14), [](auto, auto){ return false; }));
-    CHECK(pairwise_all(std::make_tuple('x'), [](auto, auto){ return true; }));
+    CHECK(pairwise_all{[](auto, auto){ return true; }}(1));
+    CHECK(pairwise_all{[](auto, auto){ return false; }}(3.14));
+    CHECK(pairwise_all{[](auto, auto){ return true; }}('x'));
   }
 
   SECTION("true for longer tuples iff true for all neighbouring pairs")
   {
-    CHECK(pairwise_all(std::make_tuple(1, 1.5), std::less{}));
-    CHECK(pairwise_all(std::make_tuple(1, 1.5, 2), std::less{}));
-    CHECK(!pairwise_all(std::make_tuple(1, 2.0, 2), std::less{}));
-    CHECK(!pairwise_all(std::make_tuple(1, 2.5, 2), std::less{}));
+    CHECK(pairwise_all{std::less{}}(1, 1.5));
+    CHECK(pairwise_all{std::less{}}(1, 1.5, 2));
+    CHECK(!pairwise_all{std::less{}}(1, 2.0, 2));
+    CHECK(!pairwise_all{std::less{}}(1, 2.5, 2));
```

Comment:
This would fail for `-1U, 1`. Should it use `cmp_less` from `hacks.h`?

Reply:
It never would have occurred to me to write something like `-1U`.  Based on how literals work, I expect this to be "the negation of `1U`".  So it should be a pretty large positive number.

So: what would it mean to "fail"?

I would expect the result from the function to be `false`, because we're comparing a large positive number to `1`.  But what result would you expect for `pairwise_all{std::less{}}(-1U, 1)`, and why?

---

To answer your other question: I'm only too happy to use `cmp_less` in the implementation.  I don't think it matters much either way for this use case, but maybe there's another use case I haven't yet thought of.

Reply:
Sorry, that should have been `-1, 1U`. `-1 < 1U` is `false`.

Reply:
That makes much more sense, thanks!

I'll add some test cases for `strictly_increasing` to cover this use case when I get back to my work computer.